FTB/Signatures/CrashInfo.py

```diff
@@ -1677,3 +1680,85 @@ def createShortSignature(self):
             return msg
 
         return "No TSan warning detected"
+
+
+class ValgrindCrashInfo(CrashInfo):
+    MSG_REGEX = re.compile(r"""
+        ==\d+==\s+(?P<msg>
+        (\d+(,\d+)*\sbytes\sin\s\d+(,\d+)*\sblocks\sare\s\w+\slost)|
+        (Argument\s'\w+'\sof\sfunction\smalloc.+)|
+        (Invalid\s\w+\sof\ssize.+)|
+        ((Invalid|Mismatched)\sfree\(\).+)|
+        (Process\sterminating\swith\sdefault\saction.+)|
+        (Source\sand\sdestination\soverlap)|
+        (Syscall\sparam.+)|
+        (.+?uninitialised\svalue.+)
+        )""", re.VERBOSE)
+
+    def __init__(self, stdout, stderr, configuration, crashData=None):
+        '''
+        Private constructor, called by L{CrashInfo.fromRawCrashData}. Do not use directly.
+        '''
+        CrashInfo.__init__(self)
+
+        if stdout is not None:
+            self.rawStdout.extend(stdout)
+
+        if stderr is not None:
+            self.rawStderr.extend(stderr)
+
+        if crashData is not None:
+            self.rawCrashData.extend(crashData)
+
+        self.configuration = configuration
+
+        # If crashData is given, use that to find the Valgrind trace, otherwise use stderr
+        vgdOutput = crashData if crashData else stderr
+        stackPattern = re.compile(r"""
+            ^==\d+==\s+(at|by)\s+            # find beginning of line
+            0x[0-9A-Fa-f]+\:\s+              # address
+            (?P<func>.+)\s+                  # function name
+            \((in\s+)?(?P<file>.+?)(:.+?)?\) # file name
+            """, re.VERBOSE)
+
+        for traceLine in vgdOutput:
+            if not traceLine.startswith("=="):
+                # skip unrelated noise
+                continue
+            lineInfo = re.match(stackPattern, traceLine)
+            if lineInfo is None:
+                if self.backtrace:
+                    # at this point we should have the full trace
+                    # check if address info is available
+                    addr = re.match(r"^==\d+==\s+Address\s(?P<addr>0x[0-9A-Fa-f]+)\s", traceLine)
+                    if addr:
+                        self.crashAddress = int(addr.group("addr"), 16)
+                    # done parsing
```

Comment:
If I understand this code correctly (and I may well not :D), this ends the parsing after the first non-stack line has been found. So in case of reports where you have two stacks (e.g. the invalid free example in valgrind-if-01.txt), the second stack is not included? I believe for ASan we do include that stack and just append it to the first, so it is possible to match on that second stack as well. Can you maybe check that? We should try to keep that somewhat consistent, although I agree the stack concat in that particular case is not the best solution.

Reply:
The functionality we should aim for is concatenating the use, alloc and free stacks if available, correct?

Reply:
All available stacks in a single report (not the entire log of course) will now be added to the backtrace.